### ISSUE TITLE
JIT: Branchless clamp

### ISSUE THREAD

**Issue by oleg-st** [dotnet/runtime#108431]
### Description

Code:

```csharp
using System;
public class C {
    private static uint Clamp1(uint a)
    {
        return Math.Clamp(a, 5, 100);
    }
    
    private static uint Clamp2(uint a)
    {
        return a < 5 ? 5 : (a > 100 ? 100 : a);
    }
    
    private static uint Clamp3(uint a)
    {
        return a < 100 ? (a > 5 ? a : 5) : 100;
    }  
}
```

Asm:
```asm
C:Clamp1(uint):uint (FullOpts):
       push     rbp
       mov      rbp, rsp
       cmp      edi, 5
       jb       SHORT G_M15322_IG05
       mov      eax, 100
       cmp      edi, 100
       cmovbe   eax, edi
G_M15322_IG04:  ;; offset=0x0014
       pop      rbp
       ret      
G_M15322_IG05:  ;; offset=0x0016
       mov      eax, 5
       jmp      SHORT G_M15322_IG04

C:Clamp2(uint):uint (FullOpts):
       mov      eax, 5
       mov      ecx, 100
       cmp      edi, 100
       cmovbe   ecx, edi
       cmp      edi, 5
       cmovae   eax, ecx
       ret      

C:Clamp3(uint):uint (FullOpts):
       cmp      edi, 100
       jae      SHORT G_M1176_IG05
       mov      eax, 5
       cmp      edi, 5
       cmova    eax, edi
       ret      
G_M1176_IG05:  ;; offset=0x0011
       mov      eax, 100
       ret      
```

`Clamp2` is branchless, `Clamp1` and `Clamp3` have branches.

[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGAHKAJYA3bBhhNySBgFcBAOwysANtny9yAClkKG2AJRcenaj1NMA7AwCyogBYA6FirUbsaBgFZ35AAw+9ANyG3AC+wQzh/MKi4sSSMvKKTqq8pFqJugYmRuGmxJbYDAA8ngwA/KUgDK4MAHwMvj7lDX4MVfpB2aHhkYIiYhJS2knOvHTpOvrhxmY8+brFLU0VNfUezYVVHnptS52mIdw0YdRAA)

### Configuration

.NET 8/9

**Comment by sfiruch** [comment]
I investigated the performance of `Math.Abs` some months ago and saw that the peformance of the branchless version is only better when the sign is *very* unpredictable. In all practical applications I measured, the performance of `Abs` with branches was better. The `cmov` version was only better in certain microbenchmarks.

I suspect `Clamp` would behave similarly.

Did you measure the performance impact on real-world code?

**Comment by oleg-st** [comment]
1) `Math.Clamp` has two branches (value < min and value > max), one is removed (converted to `cmov`), the other is not. `Clamp2` both are removed, `Clamp3` is similar to `Math.Clamp`. 
2) `Math.Max` and `Math.Min` both are branchless in .NET8. https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-8/#branching (see `Math.Max` sample)

> Did you measure the performance impact on real-world code?

No.

**Comment by sfiruch** [comment]
FYI: Agner Fog says the following about branches vs cmov (https://www.agner.org/optimize/optimizing_assembly.pdf#page=70):

> As a rule of thumb, we can say that a conditional jump is faster than a conditional move if the code is part of a dependency chain and the prediction rate is better than 75%

**Comment by sfiruch** [comment]
OTOH https://github.com/dotnet/perf-autofiling-issues/issues/12985 has shown great perf improvements after `cmov` was introduced for .NET8. Interesting...

**Comment by tannergooding** [comment]
The cost of `cmov` is very microarchitecture dependent. One the very latest CPUs it is often always as fast or faster than a predicted `jcc` (for trivial moves where there is no real secondary computation).

On older CPUs, it can be worse than a predicted branch, especially as the predictability increases. It likewise can be worse if their is cost to computing the conditional value, the benefit reducing as the cost to compute approaches 20 cycles.

There isn't likely a huge benefit or drawback to typical real world applications regardless of which approach is used here, so we default to writing code that is easy to read, understand and maintain. If there were a significant benefit, then it would likely be better worth our time improving the JIT to recognize the more readable pattern instead.

**Comment by AndyAyersMS** [comment]
> [!NOTE]
> This comment was generated with assistance from GitHub Copilot.

Prototyped a tweak to `OptIfConversionDsc::IfConvertCheck` that treats a `GT_SELECT` operand as cheap (capping its cost at the existing threshold), which lets the outer `min` check in `Math.Clamp` also if-convert. With that change, `Clamp1`, `Clamp2`, `Clamp3`, and the `int` overload all emit the same fully branchless `cmp;cmov;cmp;cmov` sequence.

SPMI x64 across aspnet2, benchmarks.run{,_pgo}, libraries.{crossgen2,pmi}, realworld.run: 774 contexts diff, 273 size improvements / 439 regressions / 62 same, **+675 net bytes**. PerfScore: 556 method-level improvements vs 179 regressions, but with several +20% to +36% PerfScore regressions on hot Tier1 methods (e.g. `Task.get_Status`, `JsonWriterOptions.get_IndentSize`) where the original branch was highly predictable per PGO.

This matches @tannergooding's 2024 [note](https://github.com/dotnet/runtime/issues/108431#issuecomment-2393886614): the cmov-vs-branch tradeoff is microarchitecture- and prediction-rate-dependent, and unconditionally collapsing predictable branches into cmov chains hurts. A real fix would have to gate on branch-probability data (e.g. only nest-if-convert when PGO marks the outer condition as poorly predicted, or there's no profile). That's meaningfully bigger than a small heuristic tweak, so leaving the issue open with these findings rather than opening a PR.

**Comment by vladd** [comment]
In this case, shouldn’t the JIT first check the processor family/capabilities it’s compiling the code for, and apply (or don’t apply) weights/optimizations which are valid exactly for the current processor? (This would however increase the test surface a lot.)

**Comment by tannergooding** [comment]
The consideration is just that, whether its actually beneficial or not to add the additional multiplier to the test bed complexity

Most native compilers, to my knowledge, simply default to `cmov` unless they have PGO data indicating otherwise.

**Comment by MichalPetryka** [comment]
> In this case, shouldn’t the JIT first check the processor family/capabilities it’s compiling the code for, and apply (or don’t apply) weights/optimizations which are valid exactly for the current processor? (This would however increase the test surface a lot.)

The JIT has historically avoided doing per CPU family optimizations due to difficulty in testing and complications with AOT handling.

**Comment by vladd** [comment]
@MichalPetryka 
> The JIT has historically avoided doing per CPU family optimizations due to difficulty in testing and complications with AOT handling.

> Most native compilers, to my knowledge, simply default to cmov unless they have PGO data indicating otherwise.

One could, theoretically, treat a concrete processor as a precomputed PGO.

**Comment by tannergooding** [comment]
> One could, theoretically, treat a concrete processor as a precomputed PGO.

Not really, because those are fairly distinct concepts and have other downstream impact.

The consideration here is whether to do a `x = cond ? y : z` to `cmov` transform at all; while PGO influences the weighting of the branches and therefore which is hot vs cold regardless of any such transform. The applicable weight ranges for which you'd want to kick in such a transform can be influenced by microarchitecture, but that's about all it can do.

-- And without PGO data, you have to make some reasonable default assumption about likelihoods. Many compilers tend to default to 50/50 on them or give slight preference to the "fallthough" path (matching what most CPUs do). But that still just plays into the general consideration of "if given a simple `x = cond ? y : z` like expression, where `y` and `z` are "trivial", is it worth doing the transform at all.
